### PR TITLE
Updates addressing Issue#16

### PR DESCRIPTION
**Pull request by mvfki**
Here's the modification that I made to address the request I mentioned in Issue #16 
Additionally, I tried to merge the documentation manuals of `find_partition` and `find_partition_rcpp` because they share almost the same information. I might need to keep the Rcpp function non-exposed though. Please feel free to give me any kind of feedback before you accept it.

### CODE REVIEW

**Comment by evanbiederstedt** [comment]
Thanks for this @mvfki 

I'm happy to accept this if you do one small thing for me: instead of rewriting the logic for `find_partition_rcpp()`, could you create a new function called `find_partition_rcpp_with_seed()` or something like this? 

That is, have two functions: leave the old one as it was, and then create a new function (with a new name) with `nrep` and `seed`.

That strikes me as less dangerous for any dependencies we have. 

Please update the minor version and add yourself as a package author; I'll update on CRAN. 

Sound reasonable?

Best, Evan

**Comment by szhorvat** [comment]
PRNGs are not generally designed to be used in this manner, i.e. by repeated re-seeding. If you do this, you may lose good statistical properties (such as independence between the subsequent samples). I'm not sure why you are doing this since repeated seeding does not achieve anything.

**Comment by mvfki** [comment]
> My point is that (according to the description), this PR does an analog of:
> 
> ```
> sapply(1:10, function (s) { set.seed(s); runif(1) })
> ```
> 
> This is a misuse of the RNG, and depending on how the RNG works may actually break independence between the samples.
> 
> The proper way is to seed _once_ at the beginning, then generate the sample without re-seeding, not to re-seed with incrementally changed seeds after every

@szhorvat Got you. Thanks for clarifying.

and @evanbiederstedt @theAeon , how about I do something like this in the new function
```
# the R interface
find_partition_with_rep <- function(...{the same args as original}, nrep=1, seed=NULL) {
   # the same preprocessing
   if (!is.null(seed)) set.seed(seed)
   find_partition_with_rep_rcpp(...)
}
```

```
// the Rcpp function
vector<size_t> find_partition_with_rep_rcpp(..., int nrep=1) {
  // Just do not do anything like getting the seed or incrementing the seed
  // However, since the optimiser object needs a "seed" number as argument
  // I'll just do the exactly same thing as the original version does there
  igraph_t g;
  igraph_vector_t edges;
  igraph_vector_init(&edges, edgelist_length);
  Stl_To_Igraph_vector_t(edgelist, &edges);
  igraph_create(&g, &edges, num_vertices, direction);

  Graph og(&g, edge_weights);

  double best_quality = -1;
  std::vector<size_t> best_cluster;
  for(int i=0; i<nrep; i++) {
    Rcpp::checkUserInterrupt();
    Optimiser o( (int) (R::runif(0,1)*(double)RAND_MAX) );
    RBConfigurationVertexPartition p(&og,resolution);
    double val=1;
    int iter=0;
    while(val>0 && (iter<niter || niter<0)) {
      val=o.optimise_partition(&p);
      iter++;
    }
    if (p.quality(resolution) > best_quality) {
      best_cluster = p.membership();
      best_quality = p.quality(resolution);
    }
  }

  igraph_destroy(&g);
  igraph_vector_destroy(&edges);

  return(best_cluster);
}
```

**Comment by mvfki** [comment]
I guess we don't even need the seed argument for the R user interface then. People can just do `set.seed()` by themselves.

**Comment by theAeon** [comment]
Yeah, if you're using the random double as a seed proxy the argument is unnecessary.

**Comment by evanbiederstedt** [comment]
> how about I do something like this in the new function

Looks good, feel free to propose any name you wish; just something descriptive

> Yeah, if you're using the random double as a seed proxy the argument is unnecessary.

Would you be able to write a unit test for this function as well? That would be useful

File: R/communities.R [hunk 1]
```diff
@@ -55,43 +55,58 @@ leiden.community <- function(graph, resolution=1.0, n.iterations=2) {
...
+  edgelist_length <- length(edgelist)
+  num_vertices <- length(igraph::V(graph)) - 1
+  direction <- igraph::is_weighted(graph)
+  find_partition_rcpp(edgelist, edgelist_length, num_vertices, direction, edge_weights, resolution, niter)
+}
+
+#' @param nrep Number of replicate starts with random number being updated. (default=10) The result with the best quality will be returned.
+#' @rdname find_partition
+#' @export
+find_partition_with_rep <- function(graph, edge_weights, resolution=1.0, niter = 2.0, nrep = 10) {
```
Comment (evanbiederstedt) [inline R/communities.R:82]: @mvfki Please document this as we document the other functions

e.g.

```
#' Finds the optimal partition using the Leiden algorithm
#'
#' @param graph The igraph graph to define the partition on
#' @param edge_weights Vector of edge weights. In weighted graphs, a real number is assigned to each (directed or undirected) edge. For an unweighted graph, this is set to 1. Refer to igraph, weighted graphs.
#' @param resolution Integer resoluiton parameter controlling communities detected (default=1.0) Higher resolutions lead to more communities, while lower resolutions lead to fewer communities.
#' @param niter Number of iterations that the algorithm should be run for (default=2)
#' @return A vector of membership values
#' @examples 
#' library(igraph)
#' library(leidenAlg)
#' 
#' g <- make_star(10)
#' E(g)$weight <- seq(ecount(g))
#' find_partition(g, E(g)$weight)
#' 
#' @export
```

This is for CRAN

**Review by evanbiederstedt — COMMENTED** [review]
Documentation request above

Also, if you could, please create a basic unit test and example showing how `set.seed()` works here. That will be promoted to the roxygen2 documentation, which is useful for users.

File: R/communities.R (hunk 1 above)
Comment (mvfki) [reply R/communities.R:82]: The tag `@rdname` has already done the job that directs multiple functions with the same "rdname" to one manual where all of them share the argument documentation. Like what we often see with some generic functions, when we do either `?find_partition` or `?find_partition_with_rep`, we get the identical help page.

Comment (evanbiederstedt) [reply R/communities.R:82]: Yes, but
(A) the `nrep` argument is different
(B) I think the function itself is different, and you could provide an explanation to the user explaining this

Comment (mvfki) [reply R/communities.R:82]: Got you. Will do so, together with the unit tests.

File: R/communities.R [hunk 2]
```diff
@@ -76,9 +75,33 @@ find_partition <- function(graph, edge_weights, resolution=1.0, niter = 2.0) {
   find_partition_rcpp(edgelist, edgelist_length, num_vertices, direction, edge_weights, resolution, niter)
 }
 
+#' Finds the optimal partition using the Leiden algorithm with replicate starts
```
Comment (evanbiederstedt) [inline R/communities.R:78]: This looks nice. Could you add one more sentence for an explanation? The GitHub issue + PR discussion has been very informative

File: R/communities.R [hunk 3]
```diff
@@ -77,15 +77,24 @@ find_partition <- function(graph, edge_weights, resolution=1.0, niter = 2.0) {
...
+#' a community structure with exploration, rather than blindly trusting the
+#' returned result that comes with the highest quality value.
+#'
+#' The random number generator (RNG) is not re-seeded at each new start of
+#' community detection, in order to keep the independence of each replicate. To
+#' get reproducible result, users can run \code{set.seed()} before calling these
+#' functions.
+#'
+#' \code{\link{find_partition}} only performs the community detection once and
+#' the reproducibility can also be ensured with \code{set.seed()}.
```
Comment (evanbiederstedt) [inline R/communities.R:94]: @mvfki This is great, thanks!

**Comment by evanbiederstedt** [comment]
@mvfki Looks great, thanks!

I'll merge and update on CRAN

**Review by evanbiederstedt — APPROVED** [review]
works for me

**Comment by evanbiederstedt** [comment]
v1.1.1 on CRAN, thanks @mvfki !

**Pull request merged**